File: ios/ClingSync/SessionPassphrase.swift
import Foundation

// In-memory passphrase of the currently open repository, mirroring the
// "session" storage mode. It lets the browse screen open its own repository
// session right after connecting, even when the user declined to store the
// passphrase in the keychain. It lives exactly as long as the open
// repository: every successful connect sets it, every close clears it.
enum SessionPassphrase {
    private static let lock = NSLock()
    private static var entry: (repositoryID: String, passphrase: String)?

    static func set(repositoryID: String, passphrase: String) {
        lock.lock()
        defer { lock.unlock() }
        entry = (repositoryID, passphrase)
    }

    static func clear() {
        lock.lock()
        defer { lock.unlock() }
        entry = nil
    }

    static func get(repositoryID: String) -> String? {
        lock.lock()
        defer { lock.unlock() }
        guard let entry, entry.repositoryID == repositoryID else { return nil }
        return entry.passphrase
    }
}

File: android/app/src/main/java/com/clingsync/android/SessionPassphrase.kt
package com.clingsync.android

// In-memory passphrase of the currently open repository, mirroring iOS's
// "session" storage mode. It lets the browse screen open its own repository
// session right after connecting, even when the user declined (or the device
// cannot support) storing the passphrase securely. It lives exactly as long
// as the open repository: every successful connect sets it, every close
// clears it.
object SessionPassphrase {
    @Volatile
    private var entry: Pair<String, String>? = null

    fun set(
        repositoryID: String,
        passphrase: String,
    ) {
        entry = repositoryID to passphrase
    }

    fun clear() {
        entry = null
    }

    fun get(repositoryID: String): String? {
        val current = entry ?: return null
        return if (current.first == repositoryID) current.second else null
    }
}
